salt/cloud/clouds/vmware.py

```diff
@@ -2258,49 +2311,65 @@ def create(vm_):
-                folder_ref = object_ref.parent
-            else:
-                folder_ref = datacenter_ref.vmFolder
+    else:
+        clone_type = None
+        object_ref = None
+
+    # Either a cluster, or a resource pool must be specified when cloning from template or creating.
+    if resourcepool:
+        resourcepool_ref = _get_mor_by_property(vim.ResourcePool, resourcepool)
+        if not resourcepool_ref:
+            log.error("Specified resource pool: '{0}' does not exist".format(resourcepool))
+            if not clone_type or clone_type == "template":
+                raise SaltCloudSystemExit('You must specify a resource pool that exists.')
+    elif cluster:
+        cluster_ref = _get_mor_by_property(vim.ClusterComputeResource, cluster)
+        if not cluster_ref:
+            log.error("Specified cluster: '{0}' does not exist".format(cluster))
+            if not clone_type or clone_type == "template":
+                raise SaltCloudSystemExit('You must specify a cluster that exists.')
         else:
+            resourcepool_ref = cluster_ref.resourcePool
+    elif clone_type == "template":
+        raise SaltCloudSystemExit(
+            'You must either specify a cluster or a resource pool when cloning from a template.'
+        )
+    elif not clone_type:
+        raise SaltCloudSystemExit(
+            'You must either specify a cluster or a resource pool when not cloning.'
+        )
+    else:
+        log.debug("Using resource pool used by the {0} {1}".format(clone_type, vm_['clonefrom']))
+
+    log.trace('resourcepool_ref: {0}'.format(resourcepool_ref))
+
+    # Either a datacenter or a folder can be optionally specified when cloning, required when creating.
+    # If not specified when cloning, the existing VM/template\'s parent folder is used.
+    if folder:
+        folder_ref = _get_mor_by_property(vim.Folder, folder)
+        if not folder_ref:
+            log.error("Specified folder: '{0}' does not exist".format(folder))
             log.debug("Using folder in which {0} {1} is present".format(clone_type, vm_['clonefrom']))
             folder_ref = object_ref.parent
+    elif datacenter:
+        datacenter_ref = _get_mor_by_property(vim.Datacenter, datacenter)
+        if not datacenter_ref:
+            log.error("Specified datacenter: '{0}' does not exist".format(datacenter))
+            log.debug("Using datacenter folder in which {0} {1} is present".format(clone_type, vm_['clonefrom']))
+            folder_ref = object_ref.parent
+        else:
+            folder_ref = datacenter_ref.vmFolder
+    elif not clone_type:
+        raise SaltCloudSystemExit(
+            'You must either specify a folder or a datacenter when creating not cloning.'
+        )
+    else:
+        log.debug("Using folder in which {0} {1} is present".format(clone_type, vm_['clonefrom']))
+        folder_ref = object_ref.parent
 
+    log.trace('folder_ref: {0}'.format(folder_ref))
```

Comment:
This should be removed. All the data once prepared gets dumped in the debug log anyway.